### ISSUE TITLE
Converting the Sharp/Node Duplex to web standard ReadableStream

### ISSUE THREAD

**Issue by FunctionDJ** [lovell/sharp#4234]
### What are you trying to achieve?

I'm trying to convert a `Sharp` object as returned by `sharp().resize(...)` to a [web standard `ReadableStream<Uint8Array>`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream) for compatibility with [Deno.serve()](https://docs.deno.com/api/deno/~/Deno.serve) / [`new Response(body)`](https://developer.mozilla.org/en-US/docs/Web/API/Response/Response#parameters)

### When you searched for similar issues, what did you find that might be related?

I have searched a lot, e.g. in the "issues" tab on this repo, but couldn't find any truly related issue. (Aside from #4013 maybe)
I managed to make it work with:
- `new ReadableStream({ start(controller) { sharpObject.on("data", chunk => controller.enqueue(chunk) } })` - but it's slightly slower than the classic `sharpObject.pipe(res)` you'd use with `http.createServer` or express
- `Readable.toWeb(sharpObject)` - appears to pipe some image data, but the served image appears to be broken/blank and it's significantly slower (~3x) on a 20MB input file than `.pipe(res)` or the previous `new ReadableStream` option
*(Note: `toWeb` was added in Node v17 and is still marked as experimental)*

Somewhat related: Stackoverflow posts about the difference between web standard ReadableStream and Node Readable
- https://stackoverflow.com/a/66629140/9948553
- https://stackoverflow.com/q/37614649/9948553

### Please provide a minimal, standalone code sample, without other dependencies, that demonstrates this question

```ts
const sharpObject = sharp("/home/function/Pictures/img.png").webp();
const readableStream: ReadableStream<Uint8Array> = sharpObject;
/**
 * TypeScript will already warn here that
 * ` 'Sharp' is missing the following properties from type 'ReadableStream<Uint8Array>' `,
 * which at runtime holds true as the `readableStream` variable doesn't work
 * with e.g. `new Response(readableStream)` in Deno
 */
```

### Please provide sample image(s) that help explain this question

independent of image file

**Comment by lovell** [comment]
> `Readable.toWeb(sharpObject)` - appears to pipe some image data, but the served image appears to be broken/blank

Use of `toWeb` is probably what I'd suggest people try first when converting. Are you able to provide a minimal code sample that allows someone else to reproduce the problem you ran into?

**Comment by FunctionDJ** [comment]
@lovell 
This first example works both when run with Deno and with Node and has equal performance on average in my test case:
```ts
import http from "node:http";
import sharp from "sharp";

http
	.createServer((_req, res) => {
		const sharpObject = sharp("img.png").webp();
		sharpObject.pipe(res);
	})
	.listen(8000);
```

This second example only works in Deno and is about 1% slower on average when converting a large, 20MB file, only 1 request at a time, i did not test multiple concurrent requests yet. More importantly, the response appears to be incomplete. Chromium will log `GET http://localhost:8000/ net::ERR_INCOMPLETE_CHUNKED_ENCODING 200 (OK)` in the console and the bottom ~15% of the image appear as gray and Postman reports ~800KB response size vs. the 1MB response of the first example.

```ts
import sharp from "npm:sharp";
import { Readable } from "node:stream";

Deno.serve(() => {
	const sharpObject = sharp("img.png").webp();
	return new Response(Readable.toWeb(sharpObject) as ReadableStream);
});
```

**Comment by lovell** [comment]
This looks a bit like the problem reported at https://github.com/denoland/deno/issues/24606

If performance is of concern and memory is not a limiting factor, perhaps experiment with Buffers rather than Streams to avoid many small memory allocations.

**Comment by FunctionDJ** [comment]
I found that Deno issue too, but i'm not using Fresh or http2 (intentionally).

As for Buffers, i'm not aware that they can be streamed. If i use `sharpObject.toBuffer()`, then the code will wait for the conversion to finish, keeping all data in memory for the moment (if i understood correctly).

**Comment by ducan-ne** [comment]
Bun is working with the following code
```typescript
import sharp from "sharp"
import { Readable } from "node:stream"
import { Buffer } from "node:buffer"

Bun.serve({
  fetch: async (req) => {
    const svg = await req.text()
    const sharpObject = sharp(Buffer.from(svg)).webp()
    return new Response(Readable.toWeb(sharpObject) as ReadableStream, {
      headers: {
        'Content-Type': 'image/webp',
      },
    })
  },
})

```

But it getting this error message for each request, not sure why it happens
```
error: Premature close
 code: "ERR_STREAM_PREMATURE_CLOSE"

      at new NodeError (node:stream:244:20)
      at node:stream:714:47
      at emit (node:events:48:48)
      at emit (node:events:48:48)
      at endReadableNT (node:stream:2207:27)
      ```

**Comment by FunctionDJ** [comment]
@lovell No. The project that caused me to write this issue is in hiatus and i might rewrite it in Go for performance reasons that are unrelated to sharp - aside from this particular issue in combination with Deno.

If you're asking because you'd like to close abandoned issues, it's your project, but i think "native" WHATWG compatibility might come up again.

**Comment by Eptagone** [comment]
I'm getting a similar error to the one mentioned by @ducan-ne. But in my case I'm using NodeJS

```ts
import sharp from "sharp";

function generateStreamResponse(buffer: ArrayBuffer) {
    const operations = sharp(buffer)
        .toFormat("webp");
    const stream = Readable.toWeb(operations) as ReadableStream;
    return Response(stream, { headers: { "Content-Type": "image/webp" } });
}
```

```log
  Stack trace:
    at /Users/user/source/my-project/node_modules/.pnpm/sharp@0.34.5/node_modules/sharp/lib/output.js:1585:35
    [...] See full stack trace in the browser, or rerun with --verbose.
  Caused by:
  Premature close
    at Sharp.onclose (node:internal/streams/end-of-stream:171:30)
    [...] See full stack trace in the browser, or rerun with --verbose.
```

However, I just noticed that it works fine when using ReadableStream (node) as input like this:

```ts
import sharp from "sharp";
import type { ReadableStream as NodeReadableStream } from "stream/web";

function generateStreamResponse(input: NodeReadableStream) {
    const operations = sharp()
        .toFormat("webp");
    const stream = Readable.toWeb(
        Readable.fromWeb(input).pipe(operations)
    ) as ReadableStream;
    return Response(stream, { headers: { "Content-Type": "image/webp" } });
}
```

**Comment by lovell** [comment]
I wonder if this might be a duplex vs half-duplex issue? Does the following help (requires Node.js v24.14.0+)?
```js
const { readable, writable } = Duplex.toWeb(operations, { type: 'bytes' });
```
https://nodejs.org/api/stream.html#streamduplextowebstreamduplex-options

**Comment by Eptagone** [comment]
> I wonder if this might be a duplex vs half-duplex issue? Does the following help (requires Node.js v24.14.0+)?
> 
> const { readable, writable } = Duplex.toWeb(operations, { type: 'bytes' });
> https://nodejs.org/api/stream.html#streamduplextowebstreamduplex-options

I think it works! Thx.

The only problem I have now is a typing issue because **@types/node** for Node 24 is still with v24.12.4 types and the [DuplexToWebOptions](https://nodejs.org/docs/v24.15.0/api/stream.html#streamduplextowebstreamduplex-options) were added since v24.14.0. Also looks like the `options.type` is now `options.readableType` after node v24.15.0. But we can use @ts-ignore until they add the missing types.

**Comment by lovell** [comment]
@Eptagone Thanks for confirming, I've added an example to the docs via commit https://github.com/lovell/sharp/commit/792c2941c3e12e0dd76f34d626f2b0011d6b6c7b

I'll leave this open as a future possible enhancement to add first-class support for Web Streams.

For now, anyone else experiencing this should use `Duplex.toWeb` for conversion.

**Comment by Eptagone** [comment]
Thx @lovell. The `await` for fetch in the example is missing by the way.

**Comment by Huaian666** [comment]
exactly what i needed rn, thanks for building it

**Comment by Huaian666** [comment]
would love this. right now i have to use a separate tool just for this one thing
